DigitalLearningSolutions.Data.Tests/Services/PasswordServiceTests.cs

```diff
@@ -48,5 +51,127 @@ public async Task Changing_password_does_not_save_plain_password()
+            A.CallTo(() => passwordDataService.SetPasswordByUserReferenceAsync(A<UserReference>._, A<string>._))
+                .MustNotHaveHappened();
+        }
+
+        [Test]
+        public async Task Changing_password_for_linked_accounts_uses_admin_email_if_no_delegate_email()
+        {
+            // Given
+            var admin = Builder<AdminUser>.CreateNew().With(u => u.EmailAddress = "email").Build();
+            A.CallTo(() => cryptoService.GetPasswordHash("new-password")).Returns("hash-of-password");
+
+            // When
+            await passwordService.ChangePasswordForLinkedUserAccounts(admin, null, "new-password");
+
+            // Then
+            A.CallTo(() => passwordDataService.SetPasswordByEmailAsync("email", "hash-of-password"))
+                .MustHaveHappened(1, Times.Exactly);
+            A.CallTo(() => passwordDataService.SetPasswordByUserReferenceAsync(A<UserReference>._, A<string>._))
+                .MustNotHaveHappened();
+        }
+
+        [Test]
+        public async Task Changing_password_for_linked_accounts_uses_delegate_email_if_no_admin_email()
+        {
+            // Given
+            var candidate = Builder<DelegateUser>.CreateNew().With(u => u.EmailAddress = "email").Build();
+            A.CallTo(() => cryptoService.GetPasswordHash("new-password")).Returns("hash-of-password");
+
+            // When
+            await passwordService.ChangePasswordForLinkedUserAccounts(null, candidate, "new-password");
+
+            // Then
+            A.CallTo(() => passwordDataService.SetPasswordByEmailAsync("email", "hash-of-password"))
+                .MustHaveHappened(1, Times.Exactly);
+            A.CallTo(() => passwordDataService.SetPasswordByUserReferenceAsync(A<UserReference>._, A<string>._))
+                .MustNotHaveHappened();
+        }
+
+        [Test]
+        public async Task Changing_password_for_linked_accounts_uses_user_ids_if_no_email()
+        {
+            // Given
+            var admin = Builder<AdminUser>.CreateNew().With(u => u.EmailAddress = null).With(u => u.Id = 34).Build();
+            var candidate = Builder<DelegateUser>.CreateNew().With(u => u.EmailAddress = null).With(u => u.Id = 309)
+                .Build();
+            A.CallTo(() => cryptoService.GetPasswordHash("new-password")).Returns("hash-of-password");
+
+            // When
+            await passwordService.ChangePasswordForLinkedUserAccounts(admin, candidate, "new-password");
+
+            // Then
+            A.CallTo(() => passwordDataService.SetPasswordByEmailAsync(A<string>._, A<string>._))
+                .MustNotHaveHappened();
+            A.CallTo(
+                    () => passwordDataService.SetPasswordByUserReferenceAsync(
+                        new UserReference(34, UserType.AdminUser),
+                        "hash-of-password"
+                    )
+                )
+                .MustHaveHappened(1, Times.Exactly);

```

Comment:
This is a bit clunky to read in the test, so I'm wondering whether it's worth pulling this assertion into private method and doing something like:
`AssertPasswordSetForUserOnce(userRef, hash)`
in the tests.. 🤔